From d45f6e9c34f5b00d3a6567907904794d2b328af2 Mon Sep 17 00:00:00 2001
From: Emergi Idan <idan.emergi@intel.com>
Date: Tue, 22 Sep 2020 20:57:07 +0300
Subject: [PATCH] LMTSYS-3975: Add clock requests for LGM PPv4

Add ppv4 clock info to device tree and enable the clock
during driver boot sequence
---
 drivers/net/ppv4/pp-dev/pp_dev.c | 65 +++++++++++++++++++++++++++++++++++-----
 drivers/net/ppv4/pp-dev/pp_dev.h |  1 -
 drivers/net/ppv4/pp-dev/pp_dts.c | 29 ------------------
 3 files changed, 58 insertions(+), 37 deletions(-)

diff --git a/drivers/net/ppv4/pp-dev/pp_dev.c b/drivers/net/ppv4/pp-dev/pp_dev.c
index a1e2fdd9f3a7..bd6475b72d08 100644
--- a/drivers/net/ppv4/pp-dev/pp_dev.c
+++ b/drivers/net/ppv4/pp-dev/pp_dev.c
@@ -45,6 +45,13 @@ static struct dentry *pp_hal_dbgfs_dir;
 /* PP sysfs directory */
 struct kobject *pp_sysfs_dir;
 
+/* PP clocks definitions based on SoC type */
+#ifdef CONFIG_PPV4_LGM
+	const char *pp_clocks[] = { "ppv4" };
+#else
+	const char *pp_clocks[] = { "ppv4", "freq" };
+#endif
+
 static const struct of_device_id pp_dev_match[] = {
 	{ .compatible = "intel,ppv4" },
 	{},
@@ -159,10 +166,52 @@ static void __pp_dbg_init(void)
 	}
 }
 
+static int init_clocks(struct platform_device *pdev)
+{
+	const char *clock_name;
+	struct clk *clk;
+	u32 i;
+
+	for (i = 0; i < ARRAY_SIZE(pp_clocks); i++) {
+		clock_name = pp_clocks[i];
+		clk = devm_clk_get(&pdev->dev, clock_name);
+		if (IS_ERR(clk)) {
+			pr_err("failed to get clock '%s', error %ld\n",
+			       clock_name, PTR_ERR(clk));
+			return -ENODEV;
+		}
+
+		pr_info("enabling pp clock '%s'\n", clock_name);
+		clk_prepare_enable(clk);
+	}
+
+	return 0;
+}
+
+static void disable_clocks(struct platform_device *pdev)
+{
+	const char *clock_name;
+	struct clk *clk;
+	u32 i;
+
+	for (i = 0; i < ARRAY_SIZE(pp_clocks); i++) {
+		clock_name = pp_clocks[i];
+		clk = devm_clk_get(&pdev->dev, clock_name);
+		if (IS_ERR(clk)) {
+			pr_err("failed to get clock '%s', error %ld\n",
+			       clock_name, PTR_ERR(clk));
+			continue;
+		}
+
+		pr_info("disabling pp clock '%s'\n", clock_name);
+		clk_disable_unprepare(clk);
+	}
+}
+
 /**
  * @brief Exit pp hal modules
  */
-static void __pp_hal_exit(void)
+static void __pp_hal_exit(struct platform_device *pdev)
 {
 	port_dist_exit();
 	rpb_exit();
@@ -174,6 +223,7 @@ static void __pp_hal_exit(void)
 	rx_dma_exit();
 	bm_exit();
 	infra_exit();
+	disable_clocks(pdev);
 }
 
 #if IS_ENABLED(PP_POWER_ON_REGS_TEST)
@@ -272,7 +322,7 @@ static void __pp_initial_hw_test(void)
  * @brief Init pp hal modules
  * @return s32 0 on success
  */
-static s32 __pp_hal_init(void)
+static s32 __pp_hal_init(struct platform_device *pdev)
 {
 	s32 ret = 0;
 
@@ -280,9 +330,10 @@ static s32 __pp_hal_init(void)
 	__pp_initial_hw_test();
 #endif
 
-	if (dts_cfg->ext_clk.valid) {
-		clk_prepare_enable(dts_cfg->ext_clk.ppv4_gate_clk);
-		clk_prepare_enable(dts_cfg->ext_clk.ppv4_freq_clk);
+	ret = init_clocks(pdev);
+	if (unlikely(ret)) {
+		pr_err("Failed to init the clocks\n");
+		goto out;
 	}
 
 	ret = pp_regs_init();
@@ -474,7 +525,7 @@ static int __pp_dev_probe(struct platform_device *pdev)
 
 	__pp_dbg_init();
 
-	probe_ret = __pp_hal_init();
+	probe_ret = __pp_hal_init(pdev);
 	if (unlikely(probe_ret)) {
 		pr_err("Failed to init pp hal drivers\n");
 		return probe_ret;
@@ -519,7 +570,7 @@ static int __pp_dev_remove(struct platform_device *pdev)
 		pput_exit();
 
 	__pp_api_exit();
-	__pp_hal_exit();
+	__pp_hal_exit(pdev);
 
 	dev_set_drvdata(&pdev->dev, NULL);
 	if (dts_cfg)
diff --git a/drivers/net/ppv4/pp-dev/pp_dev.h b/drivers/net/ppv4/pp-dev/pp_dev.h
index 0ae8cba842fe..82a4939480f1 100644
--- a/drivers/net/ppv4/pp-dev/pp_dev.h
+++ b/drivers/net/ppv4/pp-dev/pp_dev.h
@@ -49,7 +49,6 @@ struct pp_dts_cfg {
 	struct pp_bm_init_param      bm_params;
 	struct uc_init_params        uc_params;
 	struct pp_misc_init_param    misc_params;
-	struct pp_ext_clk            ext_clk;
 };
 
 s32 pp_dts_cfg_get(struct platform_device *pdev, struct pp_dts_cfg *cfg);
diff --git a/drivers/net/ppv4/pp-dev/pp_dts.c b/drivers/net/ppv4/pp-dev/pp_dts.c
index 2688414d57f7..20760827f5fb 100644
--- a/drivers/net/ppv4/pp-dev/pp_dts.c
+++ b/drivers/net/ppv4/pp-dev/pp_dts.c
@@ -11,7 +11,6 @@
 #include <linux/device.h>
 #include <linux/kernel.h>
 #include <linux/errno.h>
-#include <linux/clk.h>
 #include <linux/pp_api.h>
 #include <linux/pp_qos_api.h>
 #include "pp_dev.h"
@@ -841,16 +840,6 @@ s32 __dts_cfg_get(struct pp_dts_cfg *dts_cfg)
 
 void pp_dts_cfg_dump(struct pp_dts_cfg *cfg)
 {
-	if (cfg->ext_clk.ppv4_freq_clk) {
-		pr_debug("PPv4 External Clock %#lx\n",
-			 (ulong)cfg->ext_clk.ppv4_freq_clk);
-	}
-
-	if (cfg->ext_clk.ppv4_gate_clk) {
-		pr_debug("PPv4 Gate Clock %#lx\n",
-			 (ulong)cfg->ext_clk.ppv4_gate_clk);
-	}
-
 #ifdef CONFIG_PPV4_LGM
 	pr_info("Num sessions %u\n", cfg->cls_param.num_sessions);
 	pr_info("Num syncqs   %u\n", cfg->smgr_params.num_syncqs);
@@ -1022,29 +1011,11 @@ s32 pp_dts_cfg_get(struct platform_device *pdev, struct pp_dts_cfg *cfg)
 		 cfg->misc_params.ppv4_regs_sz);
 	cfg->misc_params.valid = true;
 
-	/* Fetch external clock */
-	cfg->ext_clk.ppv4_freq_clk = devm_clk_get(&pdev->dev, "freq");
-	if (IS_ERR(cfg->ext_clk.ppv4_freq_clk))
-		cfg->ext_clk.ppv4_freq_clk = NULL;
-
-	cfg->ext_clk.ppv4_gate_clk = devm_clk_get(&pdev->dev, "ppv4");
-	if (IS_ERR(cfg->ext_clk.ppv4_gate_clk))
-		cfg->ext_clk.ppv4_gate_clk = NULL;
-
-	if (cfg->ext_clk.ppv4_freq_clk && cfg->ext_clk.ppv4_gate_clk)
-		cfg->ext_clk.valid = true;
-
 	ret = __dts_cfg_get(cfg);
 	if (ret) {
 		dev_err(&pdev->dev, "__dts_cfg_get failed (%d)\n", ret);
 		return -ENODEV;
 	}
 
-	/* Verify we have a clock source */
-	if (!cfg->ext_clk.valid && !cfg->infra_params.valid) {
-		pr_err("No clock source available\n");
-		return -ENODEV;
-	}
-
 	return ret;
 }
